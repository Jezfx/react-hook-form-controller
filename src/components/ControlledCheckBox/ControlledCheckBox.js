import React, { useMemo } from "react";
import PropTypes from "prop-types";
import { Controller, useFormContext } from "react-hook-form";
import {
  Checkbox,
  FormControlLabel,
  FormControl,
  FormHelperText,
  FormLabel,
  FormGroup,
} from "@material-ui/core";

import { isStringIncludedInArray } from "./ControlledCheckBox.utils";

const ControlledCheckBox = ({ name, label, options = [] }) => {
  const { control, errors, watch } = useFormContext();
  const error = errors[name];
  const currentValues = watch(name);
  const isInCurrentValues = isStringIncludedInArray(currentValues);

  const renderOptions = useMemo(
    () =>
      options?.map((option, index) => (
        <FormControlLabel
          key={`${name}[${index}]`}
          label={option.label}
          control={
            <Controller
              name={`${name}[${index}]`}
              control={control}
              render={({ onChange, name }) => (
                <Checkbox
                  name={`${name}[${index}]`}
                  value={`${name}[${option.value}]`}
                  checked={isInCurrentValues(option.value)}
                  onChange={(e) => {
                    e.target.checked ? onChange(option.value) : onChange();
                  }}
                />
              )}
            />
          }
        />
      )),
    [control, isInCurrentValues, name, options]
  );

  return (
    <FormControl component="fieldset" error={!!error}>
      <FormLabel component="legend">{label}</FormLabel>
      <FormGroup>{renderOptions}</FormGroup>
      <FormHelperText>{error?.message}</FormHelperText>
    </FormControl>
  );
};

ControlledCheckBox.propTypes = {
  group: PropTypes.shape({
    key: PropTypes.string,
    options: PropTypes.arrayOf(PropTypes.object),
  }),
  name: PropTypes.string,
};

ControlledCheckBox.defaultProps = {
  group: {
    key: "",
    options: [],
  },
  name: "",
};

export default ControlledCheckBox;
