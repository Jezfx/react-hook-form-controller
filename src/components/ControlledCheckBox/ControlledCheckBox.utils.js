import { curry, flow, map, toLower, includes } from "lodash/fp";

export const isStringIncludedInArray = curry((currentValues, value) =>
  flow(map(toLower), includes(toLower(value)))(currentValues)
);
