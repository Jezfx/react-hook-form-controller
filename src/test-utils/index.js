import { render } from "@testing-library/react";
import { useForm, FormProvider } from "react-hook-form";

export const renderWithProviders = (ui, { defaultValues = {} } = {}) => {
  const Wrapper = ({ children }) => {
    const methods = useForm({});
    return <FormProvider {...methods}>{children}</FormProvider>;
  };
  return render(ui, { wrapper: Wrapper });
};
