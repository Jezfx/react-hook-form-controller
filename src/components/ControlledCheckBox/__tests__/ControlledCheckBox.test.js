import React from "react";
import ControlledCheckBox from "../";
import { renderWithProviders } from "../../../test-utils";

describe("ControlledCheckBox", () => {
  it("should match snapshot", () => {
    const { asFragment } = renderWithProviders(<ControlledCheckBox />);

    expect(asFragment()).toMatchSnapshot();
  });
});
