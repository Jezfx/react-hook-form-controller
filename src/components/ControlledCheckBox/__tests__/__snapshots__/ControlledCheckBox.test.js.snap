// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledCheckBox should match snapshot 1`] = `
<DocumentFragment>
  <fieldset
    class="MuiFormControl-root"
  >
    <legend
      class="MuiFormLabel-root"
    />
    <div
      class="MuiFormGroup-root"
    />
    <p
      class="MuiFormHelperText-root"
    />
  </fieldset>
</DocumentFragment>
`;
